sprocs/assessment_instances_revoke_reservation_access.sql

```diff
@@ -0,0 +1,30 @@
+CREATE FUNCTION
+    assessment_instances_revoke_reservation_access(
+        IN arg_assessment_instance_id bigint,
+        IN arg_authn_user_id bigint
+    ) RETURNS VOID
+AS $$
+BEGIN
+    PERFORM 1
+    FROM
+       (SELECT DISTINCT r.id
+        FROM
+            assessment_instances AS ai
+            JOIN assessment_access_rules AS aar ON (aar.assessment_id = ai.assessment_id AND exam_uuid IS NOT NULL)
+            JOIN pt_exams AS x ON (x.uuid = aar.exam_uuid)
+            LEFT JOIN groups AS g ON (g.id = ai.group_id)
+            LEFT JOIN group_users AS gu ON (gu.group_id = g.id)
+            JOIN users AS u ON (u.user_id = ai.user_id OR u.user_id = gu.user_id)
+            JOIN pt_enrollments AS e ON (e.user_id = u.user_id AND e.course_id = x.course_id)
+            JOIN pt_reservations AS r ON (r.enrollment_id = e.id AND r.exam_id = x.id)
+        WHERE
+            ai.id = arg_assessment_instance_id
+            AND g.deleted_at IS NULL
+            AND (now() BETWEEN r.access_start AND r.access_end)
+        ) AS res
+        CROSS JOIN LATERAL reservations_student_end_exam(res.id, arg_authn_user_id);
```

Comment:
I'm afraid that we can't call PT sprocs directly because they are in separate schema (see #4412). I believe this code should fail if you try it locally with both PL & PT running.

Possible alternatives:
1. Add an API endpoint in PT to receive `end_exam` calls.
2. Do the full `reservations_student_end_exam` in PL.
3. Get PT to install a public version of the `end_exam` sproc that is callable by PL. This could be a special `end_exam_for_assessment_instance_id()` or similar.
4. Something else?

Reply:
1. That may be possible, but since it's server-side may be a bit harder to achieve.
2. That would involve adding migrations for audit events, sessions (to populate the audit events), etc.
3. This may work, but probably needs more consideration. In particular, if we start working with public sprocs, there may need to be a clear distinction between sprocs that should be public and sprocs that shouldn't.